### PR TITLE
Fix some complex vjps

### PR DESCRIPTION
**Pull request by angeloskath**
It doesn't fix all of them but it fixes the vjps of multiply, divide and the ffts. At some point we should do another pass for functions like power, scan, reduce etc and add more tests as well.

### CODE REVIEW

File: tests/autograd_tests.cpp [hunk 1]
```diff
@@ -1149,7 +1149,7 @@ TEST_CASE("test complex gradients") {
     auto cotan = array(complex64_t{2.0, 3.0});
     out = vjp([x](array a) { return multiply(a, x); }, y, cotan).second;
     CHECK_EQ(out.dtype(), float32);
-    CHECK_EQ(out.item<float>(), -8.0);
```
Comment (awni) [inline tests/autograd_tests.cpp:1152]: I think when I originally wrote this I was using Jax as a reference and it gave `-8.0`. Is that a bug / any ideas about that?

```python
import jax
import jax.numpy as jnp

y = jnp.array(3.0)
x = jnp.array(2.0) + 3.0j
cotan = jnp.array(2.0) + 4.0j

def fun(y):
    return y * x

print(jax.vjp(fun, y)[1](cotan))
```

Comment (angeloskath) [reply tests/autograd_tests.cpp:1152]: So Jax computes `dl/dz` and the descent direction is `dl/dz*`.  We can choose to do the same but then we 'd have to make sure that functions like `abs` return `dl/dz` instead of `dl/dz*`. And then the user would have to conjugate any complex gradient before applying it.

To talk specifically about the example above, the difference is that Jax assumes that cotan is `dl/dz` and PyTorch (and now us) assume that the cotton is `dl/dz*`

Comment (awni) [reply tests/autograd_tests.cpp:1152]: Gotcha! Thanks for the explanation.

File: mlx/primitives.cpp [hunk 2]
```diff
@@ -1950,30 +1952,62 @@ std::vector<array> FFT::vjp(
...
+    bool odd = cotangents[0].shape(axes_.back()) % 2;
+    Shape c(in.ndim(), 1);
+    c[axes_.back()] = N;
+    array indices = reshape(arange(N, stream()), std::move(c), stream());
+    array first(0, indices.dtype());
+    array last(N - 1 + odd, indices.dtype());
+    array one(1 / n_elements, in.dtype());
+    array two(2 / n_elements, in.dtype());
+    array mask =
+        where((first < indices) & (indices < last), two, one, stream());
```
Comment (awni) [inline mlx/primitives.cpp:1976]: These ops are not on the stream (e.g. `<`, `&`)

File: mlx/primitives.cpp [hunk 3]
```diff
@@ -1950,30 +1952,62 @@ std::vector<array> FFT::vjp(
...
+    bool odd = in.shape(axes_.back()) % 2;
+    Shape c(in.ndim(), 1);
+    c[axes_.back()] = N;
+    array indices = reshape(arange(N, stream()), std::move(c), stream());
+    array first(0, indices.dtype());
+    array last(N - 1 + odd, indices.dtype());
+    array one(1, complex64);
+    array half(0.5, complex64);
+    array mask =
+        where((first < indices) & (indices < last), half, one, stream());
```
Comment (awni) [inline mlx/primitives.cpp:1996]: same

**Review by awni — APPROVED** [review]
Nice fixes! Some minor comments otherwise LGTM!

File: mlx/primitives.cpp (hunk 2 above)
Comment (angeloskath) [reply mlx/primitives.cpp:1976]: Of course... silly me... I got carried away by how good it looked :-)
